citest/json_predicate/path_predicate_result.py

```diff
@@ -0,0 +1,251 @@
+    return self.__source
+
+  @property
+  def pred(self):
+    """The PathPredicate that generated the result."""
+    return self.__pred
+
+  def __init__(self, source, pred):
+    """Constructor.
+
+    Args:
+      source: [obj] The JSON object used to collect the values.
+      pred: [ValuePredicate] The predicate used to gather the results.
+    """
+    self.__source = source
+    self.__pred = pred
+    self.__path_values = []
+    self.__path_failures = []
+    self.__invalid_candidates = []
+    self.__valid_candidates = []
+
+  def add_all_path_failures(self, failures):
+    """Adds to the list of failed results.
+
+    Args:
+      failures: [list of PredicateResult] Explains reason paths were
+          omitted before we reached the final candidates..
+    """
+    self.__path_failures.extend(failures)
+    return self
+
+  def add_path_failure(self, failure):
+    """Adds to the list of pruned paths.
+
+    Args:
+      failure: [PredicateResult] Explains the reason a path was pruned.
+    """
+    self.__path_failures.append(failure)
+    return self
+
+  def add_result_candidate(self, path_value, final_result):
+    """Adds to the list of scollected values.
+
+    Args:
+      path_value: [PathValue] A value that meets the filter criteria.
+      final_result: [ValueResult] The filtering result justifying this value.
+    """
+    candidate_result = PathPredicateResultCandidate(path_value, final_result)
+    if final_result:
+      self.__valid_candidates.append(candidate_result)
+      self.__path_values.append(path_value)
+    else:
+      self.__path_failures.append(final_result)
+      self.__invalid_candidates.append(candidate_result)
+    return self
+
+  def build(self, valid=None):
+    """Construct the result.
+
+    Returns:
+      PathPredicateResult
+    """
+    if valid is None:
+      valid = len(self.__path_values) > 0
+    return PathPredicateResult(
+        valid=valid, pred=self.__pred, source=self.__source,
+        path_failures=self.__path_failures,
+        valid_candidates=self.__valid_candidates,
+        invalid_candidates=self.__invalid_candidates)
+
+
+class HasPathPredicateResult(object):
+  """This is a hack in lieu of a more meaningful interface.
+
+  The purpose of this is to allow observation validators to
+  be able to access the actual observed objects as well as how
+  they complied with constraints. This is needed for certain types
+  of constraints. Rather than coming up with the right interface at
+  this time, we'll just mooch off the PathPredicateResult which is
+  the basis of extracting values from observations and of the
+  CardinalityPredicate for counting instances.
+  """
+  # pylint: disable=too-few-public-methods
+
+  @property
+  def path_predicate_result(self):
+    """A PathPredicateResult instance providing the applicable object values."""
+    raise NotImplementedError(self.__class__.__name__)
+
+
+class PathPredicateResult(predicate.PredicateResult, HasPathPredicateResult):
```

Comment:
I'm using HasPathPredicateResult as a marker for the time being. Think of it as a different interface that class implements. It happens to be the identity. The other classes (in future CL) are marked with this because their results contain results, and can pull it out for you.

Really HasPathPredicateResult is trying to say something else, but I'm not exactly sure what that is yet so it is simpler to just call it a PathPredicateResult rather than introduce a new concept until it more clearly identifies itself.